Implement a DirectX 11 renderer

hyblocker opened the pull request:
I've implemented a DirectX 11 backend for RmlUI. Currently this has feature parity with the SDL renderer, but I plan on extending it to match the GL3 renderer in the near future (hence I'm marking this as a draft PR).

As for backends, currently this only has an implementation with the Win32 backend. I'll also be providing a GLFW backend at a later point.

I've introduced a pair of functions pointers only in the DX11 `RenderInterface` for this backend to make it more flexible and easily consumable by end users. As the current backends only support uncompressed targa files for images, and in practice one would likely want to connect it to a texture loader or something more sophisticated such as stb_image, I defined the following functions:

```cpp
// Use these typedefs to overload the default implementation with another image loader so that you can handle more than uncompressed TGA if you wish.
// Must define both the load and free functions. See RenderInterface_DX11::
typedef void (*pfnLoadTextureRaw)(const Rml::String& filename, int* pWidth, int* pHeight, uint8_t** pData, size_t* pDataSize);
typedef void (*pfnFreeTextureRaw)(uint8_t* pData);
```

Where a user may override these as follows for example:

```cpp
// Initialize the window but don't show it yet.
HWND window_handle = InitializeWindow(data->instance_handle, name, width, height, allow_resize);
if (!window_handle)
    return false;

if (!CreateDeviceD3D(window_handle)) {
    ::CloseWindow(window_handle);
    return false;
}

// Assign STB image loader to render interface
data->render_interface.LoadTextureFromFileRaw = &LoadTexture;
data->render_interface.FreeTextureFromFileRaw = &FreeTexture;
stbi_set_unpremultiply_on_load(true);
stbi_convert_iphone_png_to_rgb(true);

data->render_interface.Init(data->d3d_resources.pd3dDevice, data->d3d_resources.pd3dDeviceContext);

data->window_handle = window_handle;
data->system_interface.SetWindow(window_handle);

// Now we are ready to show the window.
::ShowWindow(window_handle, SW_SHOW);
::SetForegroundWindow(window_handle);
::SetFocus(window_handle);
```

The renderer will fallback to the Targa image loader if the user doesn't override the loader with their own one.

For completeness, below is the entire STB image loader I've been testing with. I have converted the sample assets to PNG to verify that it works correctly.

```cpp
// Add somewhere at the top of the backend
#define STB_IMAGE_IMPLEMENTATION
#include "stb_image.h"
#include <RmlUi/Core/FileInterface.h>

// STB image based loaders
void LoadTexture(const Rml::String& filename, int* pWidth, int* pHeight, uint8_t** pData, size_t* pDataSize);
void FreeTexture(uint8_t* pData);

// STB loader implementation

void LoadTexture(const Rml::String& filename, int* pWidth, int* pHeight, uint8_t** pData, size_t* pDataSize)
{
    Rml::FileInterface* file_interface = Rml::GetFileInterface();
    Rml::FileHandle file_handle = file_interface->Open(filename);
    if (!file_handle)
    {
        *pData = nullptr;
        return;
    }

    int texture_width, texture_height, num_channels;
    *pData = stbi_load_from_file((FILE*)file_handle, &texture_width, &texture_height, &num_channels, 0);

    if (*pData != nullptr)
    {
        *pWidth = texture_width;
        *pHeight = texture_height;

        // Compute number of elements in texture
        *pDataSize = texture_width * texture_height * num_channels;

        uint8_t* dataView = *pData;

        // Pre-multiply the data, otherwise it will be loaded incorrectly
        for (int i = 0; i < *pDataSize; i += 4)
        {
            dataView[i + 0] = (uint8_t)((dataView[i + 0] * dataView[i + 3]) / 255);
            dataView[i + 1] = (uint8_t)((dataView[i + 1] * dataView[i + 3]) / 255);
            dataView[i + 2] = (uint8_t)((dataView[i + 2] * dataView[i + 3]) / 255);
        }
    }
}

void FreeTexture(uint8_t* pData)
{
    stbi_image_free(pData);
}
```

hyblocker committed:
(feat): Add DX11 Win32 Backend to CMakeLists

hyblocker committed:
(feat): Implement stub DX11 backend.

Not usable in current state.

hyblocker committed:
(chore): Use spaces instead of tabs

hyblocker committed:
(feat): Make BeginFrame function take a rendetargetview

This allows users to render to a render target which may either be the swapchain (in most cases it will be) or a render texture which may then be used elsewhere (eg in a 3D scene).

hyblocker committed:
(feat): Allow users to overload the default image loader

We should provide a image loader (in this case the uncompressed TGA loader), but should also allow users using this renderer backend to overload it with their own one. E.g. a user may have libpng in their project and they want to hook libpng into the load image function to load PNGs.

hyblocker committed:
(feat): Scissor region

hyblocker committed:
(fix): Incorrect arguments into Init

hyblocker committed:
(fix): Fix double declaration

hyblocker committed:
(chore): Also require swapchain on begin frame

hyblocker committed:
(chore): Use spaces not tabs

hyblocker committed:
(chore): Snake case

hyblocker committed:
(feat): Create shaders and input layout

hyblocker committed:
(fix): Make backend free DX11 resources on shutdown

hyblocker committed:
(fix): Make input layout declaration more compact

hyblocker committed:
(fix): Free all allocated resources on Cleanup

hyblocker committed:
(fix): Reset blend state on end frame

hyblocker committed:
(fix): Use DEFAULT macros for rasterizer state

hyblocker committed:
(feat): Add SetTransform support

hyblocker committed:
(feat): Implement CBuffer

hyblocker committed:
(fix): Link to d3dcompiler

hyblocker committed:
(fix): Enable DX11 debug layer in debug mode

hyblocker committed:
(fix): Remove unnecessary padding in cbuffer

hyblocker committed:
(fix): Initialise descriptors before usage

hyblocker committed:
(fix): Feed macros into shader compiler

hyblocker committed:
(chore): Use nullptr instead of NULL

hyblocker committed:
(feat): Implement UpdateConstantBuffer

hyblocker committed:
(feat): Create texture sampler

hyblocker committed:
(feat): Implement texture loading

hyblocker committed:
(feat): Implement geometry submission

hyblocker committed:
(fix): Correct invalid input layout

hyblocker committed:
(fix): Correct incorrect rasterizer state

hyblocker committed:
(fix): Set initial state every frame

hyblocker committed:
Merge branch 'mikke89:master' into dx11-backend

hyblocker committed:
(fix): Disable sRGB for consistency with GL3

hyblocker committed:
(feat): Use a hashmap to store GPU resources

ShawnCZek commented:
It is exciting to see a renderer for DirectX 11 being implemented! I am fully aware that this is only a draft, though, as we had to implement a custom DirectX 11 renderer for our case, I would like to share one remark about the implementation.

As DirectX is mainly used for games, I would like to know if there are any plans to make some optimizations, mainly around buffers. I remember creating a new buffer for every geometry, which was quite expensive, especially when a different buffer had to be bound for every draw call. Do you plan on implementing something like transient buffers (as documented [here](https://www.gamedevs.org/uploads/efficient-buffer-management.pdf)), or are you just going to keep it simple? For our application, transient buffers brought the rendering of RmlUi to only 60% of its original time.

mikke89 commented:
I just want to say, awesome! It's great to see such solid efforts from the community :) Let me know when you feel it is ready for review, or if you would like some pointers along the way.

hyblocker commented:
> It is exciting to see a renderer for DirectX 11 being implemented! I am fully aware that this is only a draft, though, as we had to implement a custom DirectX 11 renderer for our case, I would like to share one remark about the implementation.
> 
> As DirectX is mainly used for games, I would like to know if there are any plans to make some optimizations, mainly around buffers. I remember creating a new buffer for every geometry, which was quite expensive, especially when a different buffer had to be bound for every draw call. Do you plan on implementing something like transient buffers (as documented [here](https://www.gamedevs.org/uploads/efficient-buffer-management.pdf)), or are you just going to keep it simple? For our application, transient buffers brought the rendering of RmlUi to only 60% of its original time.

Currently I'm focusing on feature parity, but transient buffers is something I'll attempt looking into in the near future.

hyblocker commented:
> I just want to say, awesome! It's great to see such solid efforts from the community :) Let me know when you feel it is ready for review, or if you would like some pointers along the way.

Regarding the stb image loader test I've been testing with, how should I go about committing that? Currently I only have the implementation locally, but I figure others would find it useful to see how to integrate a popular image loading library with RmlUI to handle more file formats than a subset of TGA files.

hyblocker committed:
(fix): Remove texture cache, release texture after generating texture view

Improves memory usage and effectively eliminates leaked objects.

hyblocker committed:
(fix): Change swapchain swap effect to flip discard

hyblocker committed:
(fix): Correct typo in swapchain_occluded

hyblocker committed:
(feat): Add depth stencil state

hyblocker committed:
(fix): Add default blend factor to blend state

hyblocker committed:
(feat): Backup and restore DX11 state on frame begin / end

hyblocker committed:
(chore): Remove redundant struct

hyblocker committed:
(chore): Use snake case

hyblocker committed:
(feat): Add support for window resizing

hyblocker committed:
(feat): Handle Aero Snap

hyblocker committed:
(chore): Whitelist DX11 backend for effects sample

mikke89 commented:
Good question. Yeah, generally it would nice to show how to integrate with image loaders, STB in particular. Ideally, we would have something like that for every backend. I don't want to add more dependencies to the backend directory, but maybe we could have some separate "backend extension" samples, or something like that. Or maybe just a snippet showing how to override the function more generally in the documentation could be sufficient.

In any case, in the mean time, I'm okay with having this in some `#ifdef`s until we figure all of that out.

hyblocker committed:
(feat): Add stb_image loader

#define RMLUI_USE_STB_IMAGE_LOADER to use it, and add stb_image.h to the include dir

hyblocker committed:
(chore): Correct code style

hyblocker committed:
(feat): Implement layer system

Note: Not finished, we still need shaders and filters and compositing

hyblocker committed:
(fix): Flip UVs for passthrough

Passthrough only appears to be used for blitting render textures on GL3, so I'll flip UVs in shader (alternatively we could flip UVs in the quad, but I'll investigate this later).

hyblocker committed:
(fix): Release depth stencil variants

hyblocker committed:
(feat): Implement clip masks

Currently broken

hyblocker committed:
(feat): Implement layer handling

hyblocker committed:
(feat): Add all shader programs

hyblocker committed:
(feat): Add all properties to cbuffer struct

Most are unused currently, will be used when UseShader is implemented.

hyblocker committed:
(fix): Correct clip masks

hyblocker committed:
(feat): Half implement layer compositing

Filters are ignored currently

hyblocker committed:
(feat): Implement shaders

hyblocker committed:
(chore): Snake case

hyblocker committed:
(feat): Implement filters (no rendering)

hyblocker committed:
(fix): Convert filter rendering to D3D11

hyblocker committed:
(fix): Bind cbuffer to fragment shader

hyblocker committed:
(fix): Correct BindTexture implementation

hyblocker committed:
(chore): Snake case

hyblocker committed:
(feat): Implement filters and blurring

hyblocker committed:
(feat): Implement debug markers

Temp to fix the rendering bugs

hyblocker committed:
(chore): Spaces not tabs

hyblocker committed:
(chore): Simplify shader structure names

hyblocker committed:
(fix): Correct blur using incorrect vertex shader

hyblocker committed:
(fix): Correct typos in shaders

hyblocker committed:
(fix): Unbind slot 1 after use

hyblocker committed:
(fix): Correct clip mask implementation

hyblocker committed:
(chore): Match comment locations with GL3

hyblocker committed:
(chore): Correct RenderGeometry having side-effects

hyblocker committed:
(fix): Bind depth stencil to match GL3 behaviour

hyblocker committed:
(fix): Don't clear alpha channel

hyblocker committed:
(chore): Make RenderLayerStack use m_ for variables

hyblocker committed:
(fix): Unbind resource views after rendering filters

hyblocker committed:
(chore): Cast bounds to float

hyblocker committed:
(fix): Fix debug layer incorrect marking event at wrong spot

hyblocker committed:
(fix): Mark cbuffer as dirty when altered

hyblocker committed:
(chore): Handle transposing matrices?

Doesn't actually work

hyblocker committed:
(feat): Begin implementing BlitRenderTarget

This is to match the behaviour of glBlitFramebuffers, as on DX11 using CopySubresourceRegion breaks rendering with layers it seems

hyblocker committed:
(feat): Add temporary render target for blitting

hyblocker committed:
(chore): Assert on invalid coords for blitting

hyblocker committed:
(feat): Add a custom blit implementation

glBlitFramebuffer doesn't have a direct equivalent, so we have to mimic it's behaviour another way

hyblocker committed:
(fix): Fix incorrect blend factor

hyblocker committed:
(fix): Correct copy to copy bounds

hyblocker committed:
(fix): Why did I commit this

hyblocker committed:
(fix): Correct scissor being set incorrectly

hyblocker committed:
(fix): Correct blit shader having incorrect maths

hyblocker committed:
(chore): Simplify binding render targets

hyblocker committed:
(fix): DrawGeometry match behaviour of GL3

hyblocker committed:
(fix): Simplify clip mask debug scope

hyblocker committed:
(chore): Remove redundant code

hyblocker committed:
(feat): Better blitting

hyblocker committed:
(fix): Correct blitting

hyblocker committed:
(fix): Correct bad blend state

hyblocker committed:
(fix): Correct color matrix implementation

Not finished. Flickering

hyblocker committed:
(fix): Unbind resources when done

hyblocker committed:
(fix): Correct binding order

Eliminates errors in the debug layer

hyblocker committed:
(fix): Correct alignment of blur shader

hyblocker committed:
(chore): Remove a lot of warnings

hyblocker committed:
(fix): Fix color matrix flickering

hyblocker committed:
(fix): Remove padding from cbuffers in shaders

hyblocker committed:
(fix): Correct bind order for MaskImage

hyblocker committed:
(fix): Fix unnamed struct warning

hyblocker committed:
(fix): Fix colour blending

hyblocker committed:
(feat): Get blur to render something?

Currently broken. Needs more work

hyblocker committed:
(fix): Fix memory leak

hyblocker committed:
(fix): Fix blur flipping image on large kernels

hyblocker committed:
(fix): Correct blur implementation

hyblocker committed:
(fix): Cleanup some DX11 resources which were being leaked

hyblocker committed:
(fix): Improve blur implementation to correct UV offset

Also add early exit in the case of a kernel of size 0

hyblocker committed:
(feat): Add early exit cases for blurring to GL3

hyblocker committed:
(fix): Blurring on specific kernels is incorrect

hyblocker committed:
(chore): Remove redundant name from type

hyblocker committed:
(feat): Implement drop shadows

hyblocker committed:
(fix): Correct blur de-jitter

hyblocker committed:
(chore): Remove debug code

hyblocker committed:
(chore): Remove redundant parameter for BeginFrame

igorsegallafa commented:
Ohh nice PR @hyblocker, I was also trying to implement DX11 renderer by myself. 
Do you know exactly what is missing to finish this PR or is there anything I could help with? Thank you!

hyblocker commented:
The PR is practically ready for merging, I'm waiting for the weekend to find the time to implement the GLFW backend and then I'll move this from draft status.

Transient buffers should still be investigated if possible but my schedule has been getting busier lately, so I think it makes more sense to merge it as a separate PR later.

hyblocker committed:
(feat): Add GLFW backend

hyblocker committed:
Merge branch 'mikke89:master' into dx11-backend

hyblocker committed:
(chore): Add DirectX11 to README

hyblocker commented:
@mikke89 The DirectX11 renderer is ready for review.

mikke89 commented:
Awesome, that's great to hear!  Very impressive, I'll try to get around to it soon.

Would you consider it feature complete compared to the OpenGL 3 renderer? Would be nice to add an entry into the feature table of the renderers in the readme.

hyblocker commented:
The DirectX 11 renderer has feature parity with the OpenGL 3 renderer, including all effects, shaders, etc. (the effects sample appears to function identically between the two). I've already added the entry to the readme in the PR (see https://github.com/mikke89/RmlUi/pull/675/commits/8ed475311f2a6c5bf9b0541a17be4113d2b2c217). It also comes with a GLFW and Win32 backend.

I've been using the DirectX11 renderer for my own project for around a month and it appears to be stable and function well.

wh1t3lord commented:
I will say some words about this implementation and here they're: 

1) There's no point in m_geometry_cache and you can just use some struct for defining your GeometryHandle and have fields for VB and IB native pointers of DirectX 11 types, you just allocate it with new operator.

It is a very bad practice using .find on render callings. Searching through hash map container without any specific purpose is still slow operation, on real-time rendering applications hash-maps used not often by frequency calling otherwise it will reduce your performance on much complex scenes immediately.

You won't even get invalid buffer if you get just don't render it but instead you do .find operation every frame, why? All information about buffers stored and provided by Rml. 

See: https://github.com/mikke89/RmlUi/blob/master/Backends/RmlUi_Renderer_GL3.cpp#L1023

2) You have really naive implementation of buffer allocation. Why do not to use one buffer for vertices & indices but sub-allocating them with some offset-allocator? Some people in this thread suggesting some transient buffers (for what?). In terms of handling resources for GPU it is better to create something once big and re-use it not even deleting (deletion when RmlUI is going to shutdown) in your case it is important to bind one buffer for vertices and indices but specify offset where your data located for current GeometryHandle. 

So you allocate single buffer on init stage say like 1 Mb/512 Kb buffer that's for VB and IB and try to use offset-allocator like this https://github.com/sebbbi/OffsetAllocator (I suggest it because I will use it in my implementations)

If you want to write your own you can try why not or you can implement Transient buffers, it will improve performance but at some point using one buffer that stores all things could be a little better I didn't compare a such implementation, but I think it could be just equal. 

As I understood you're a beginner and it is fine that you have some mistakes 😃, but I expect that you will fix them. Your implementation for many people exists as 'reference' implementation and it is supposed (people expect that at some point) the implementation is done well but in reality if we accept it now after some time some people from community will ask these questions what I state here and will ask: "why did you A instead of B, B is effective and please fix it". I am pretty sure that it might happen so I don't have any choice for staying in shadows and ignoring this review. 

For you it is improving your understanding of real-time graphics :D, for me I did a review and I do them not often tbh. 

If you still have some questions or you disagree with my opinions, feel free to write your thoughts.

hyblocker committed:
(fix): Improve MSAA sampling pattern

hyblocker committed:
(fix): Remove m_geometry_cache

As per suggestion 1 from  wh1t3lord

hyblocker commented:
I've removed m_geometry_cache, I'll try improving the buffer allocation later when I have some more free time.

mikke89 reviewed (commented):
First of all, I am really impressed. The results look very solid overall, and it's no small feat to get all of this working, especially with all of our new renderer features too.

I haven't taken a detailed dive into the implementation itself, I want to leave you some ownership over this code as well. It would be great if you can help maintain it a bit occasionally in the future, but no obligations required. Looks like it follows the overall structure of the GL3 renderer, which is very nice and helps maintain them together.

I did some quick benchmarks on our benchmark sample:

| Renderer | Idle FPS | Active FPS
| -- | -- | --
| GL2 | 1400 | 83
| GL3 | 800  | 84
| DX11 | 1500 | 82
| VK | 1200 | 134

Overall, it's on par with our other renderers. Performance-wise, it is not problematic in my view. It would be nice with some improved allocation techniques as mentioned above, I think that could help with the "active FPS" here. While I would encourage that, it's not a blocker for a first iteration in my view. And we probably want to improve that for our GL3 renderer too.

I implemented the screenshot feature locally and ran a comparison against the GL3 renderer on all our visual tests. It was pixel-perfect equal in 81/102 tests. Most of the remaining ones are only slightly pixel-different along rounded edges and for color interpolation, things that one would expect to be different between renderers. So that's a solid job!

There are a few tests I would like to highlight:

|   #   | similarity scores (%) | max pixel difference  | filename | Comment
|   --   | -- | --  | -- | --
|   13    | 30.5%    | 86   | filter_blur_orb.rml             | Looks good statically, but there's quite a lot of flickering when moving the blur orb. This is very understandable, and quite tricky to get stable. I remember spending a lot of effort with this on the GL3 renderer.
|   45    | 29.3%   | 238   | ink_overflow_bounding_box.rml   | Colors are slightly more saturated for some reason?
|   58    | 44.9%   | 136   | mask_image.rml                  | The heart image is slightly more blurry, as if the vertex positions or UV coords are just slightly off.
|   84    | 13.5%   | 528   | shader_conic_gradient.rml       | Looks like the directions are rotated, see the screenshot below.

Out of these, I would only consider the conic gradient as an actual issue. The other ones are reasonable in a first iteration, not considered blockers, but if you feel like taking a closer look then that would be nice.

Could you format the files according to our .clang-format and .editorconfig? It looks kind of close to our clang-format, but also slightly off, and the indentations are not correct.

Again, great effort! I'm very happy to see some DirectX love, so thank you for that. Let me know if there is anything I can clarify or help out with.

---

![image](https://github.com/user-attachments/assets/950ad42c-838e-45b4-9857-224f0ecccf9b)

mikke89 commented:
I managed to get these error messages somehow:

![image](https://github.com/user-attachments/assets/6d787a7d-67f0-471d-b73a-4400aa4d7fd5)

Not sure exactly how, and haven't been able to reproduce it yet. I was messing around with this sample and its settings, and doing some window resizing, minimizations and such.

hyblocker committed:
(fix): Move stb_image sample to backend.

hyblocker commented on Backends/RmlUi_Renderer_DX11.h:
```diff
@@ -0,0 +1,346 @@
+    #define NUM_MSAA_SAMPLES 2
+#endif
+
+// Use these typedefs to overload the default implementation with another image loader so that you can handle more than uncompressed TGA if you wish.
+// Must define both the load and free functions. See RenderInterface_DX11::
+typedef void (*pfnLoadTextureRaw)(const Rml::String& filename, int* pWidth, int* pHeight, uint8_t** pData, size_t* pDataSize);
+typedef void (*pfnFreeTextureRaw)(uint8_t* pData);
```
Moved it to the backend files.

hyblocker commented:
I'll tackle the issues you've found throughout the week. If I have the time I'll also investigate the tests with poor accuracy to see if I could improve them.

ShawnCZek commented on Backends/RmlUi_Renderer_DX11.h:
```diff
@@ -0,0 +1,346 @@
+    #define RMLUI_DX_ASSERTMSG(statement, msg) static_cast<void>(statement)
+#endif
+
+// Helper for a common cleanup pattern in DX11
+#define DX_CLEANUP_RESOURCE_IF_CREATED(resource)    \
+        if (resource)                               \
+        {                                           \
+            resource->Release();                    \
+            resource = nullptr;                     \
+        }
```
If we are okay with including other headers in this header file, I would suggest using `CComPtr<>` from `atlbase.h`; smart pointers over raw pointers would definitely be an improvement to avoid any possible memory leakage.

mikke89 replied:
If the headers are commonly available, that should be fine. This is not one of those microsoft language extensions is it, does it require any special compiler flags? Is it compatible with MinGW & co?

ShawnCZek replied:
Right, MinGW might become a problem. I just feel like such a smart pointer (or any way to clean up the resources, really) would be a neater way to handle them, given how many times the helper is used.

mikke89 commented:
Hey, just checking in, how is it going with this feature? I'd love to see this one merged into the library. Let me know if there is anything I can help out with :)

hyblocker commented:
Hey, I have been a little busy with some IRL things and not had the time to address the issues that were brought up. In around a month I think I should have some time to address them.

ShawnCZek commented on Backends/RmlUi_Renderer_DX11.h:
```diff
@@ -0,0 +1,334 @@
+    void Cleanup ();
+
+    // Sets up DirectX11 states for taking rendering commands from RmlUi.
+    void BeginFrame(ID3D11RenderTargetView* p_render_target_view);
```
I believe this needs to be better documented, or this parameter should be removed entirely. The render target view is treated as if it is "owned" by RmlUi; it is completely cleared in the implementation. I understand this parameter exists to handle resizing the viewport and, therefore, creating a new render target view. However, an argument could be made that the renderer should/could handle that, considering this is quite specific for DirectX.

I mention this because the OpenGL backend does not have anything like this; it only clears its framebuffers. This quite tricked me up when I tried to use your implementation.

mikke89 replied:
I am making some changes around this: Removed this render target parameter from this function, made a separate Clear function which takes this parameter instead. Removed this state altogether from the class.

hyblocker committed:
(fix): Correct conic gradients being rotated incorrectly

hyblocker committed:
(feat): Add DX11 to other feature matrix

hyblocker committed:
(fix): Use RMLUI_DX_DEBUG instead of RMLUI_DEBUG to control debug device

hyblocker committed:
(chore): Give error instead of not including the file at all

hyblocker committed:
(fix): Incorrectly had DX debugging enabled

hyblocker committed:
(chore): Fix half-written comment

hyblocker commented on Backends/RmlUi_Renderer_DX11.cpp:
```diff
@@ -0,0 +1,2990 @@
+
+enum class ShaderGradientFunction { Linear, Radial, Conic, RepeatingLinear, RepeatingRadial, RepeatingConic }; // Must match shader definitions below.
+
+// We need to round up at compile-time so that we can embed the 
```
Thought I had written the entire thing, addressed it now.

hyblocker commented on Backends/RmlUi_Renderer_DX11.h:
```diff
@@ -0,0 +1,346 @@
+    #define RMLUI_DX_ASSERTMSG(statement, msg) RMLUI_ASSERTMSG(SUCCEEDED(statement), msg)
+
+    // Uncomment the following line to enable additional DirectX debugging.
+    #define RMLUI_DX_DEBUG
```
Yes, accidentally committed it as uncommented haha.

hyblocker commented on Backends/RmlUi_Renderer_DX11.cpp:
```diff
@@ -0,0 +1,2990 @@
+
+#include <RmlUi/Core/Platform.h>
+
+#if defined RMLUI_PLATFORM_WIN32
```
Made it present an error instead.

ShawnCZek commented:
After using this backend for some time and playing around with it, I have found a very interesting edge case that does not happen with other backends (at least not with OpenGL from my testing).

If a (backdrop) filter is used and another element is rendered at the same position, the element on top is flickering like crazy. I have modified one of the samples to replicate this (**do NOT run this if you are prone to epileptic shocks**): https://github.com/ShawnCZek/RmlUi/commit/7919495bf946ccb6580429278ba971d8ab7c2c07

I assume this is an issue with the entire shared buffer. From my understanding, you discard all previous data if you map a buffer with the `D3D11_MAP_WRITE_DISCARD` flag ([reference](https://learn.microsoft.com/en-us/windows/win32/api/d3d11/ne-d3d11-d3d11_map#common-usage-of-d3d11-map-write-discard-with-d3d11-map-write-no-overwrite)). This happens, for instance, when rendering a texture for the blur effect; only the blur properties (without transform and translation) are set.

I believe there are two solutions:

1. When mapping the shared buffer, the transform and translation must also be updated. Ideally, this should happen in one place so you do not have to repeat yourself over and over again.
2. Do not use a shared buffer but multiple ones instead. While the idea of one shared constant buffer seems pretty cool, I do not think it will work without reading the previous content, which would bring a performance penalty.

By the way, I am confused as to why the `DropShadow` structure is part of the `ShaderCbuffer` union when it is a separate buffer. I guess it was a part of the shared buffer until it was also separated.

hyblocker commented:
> @hyblocker hi! A long time passed and will you continue your backend to finish? Because I can point on some things that's better to refactor :D

yeah id like to finish the backend properly

wh1t3lord commented:
@hyblocker Hi again, hope you're doing great and ready for battle. Anyway, I did some review that you really should to implement buffer manager and persistent buffers (I wrote review comment about it). 

And you did the big job so thanks for DirectX 11 and a new backend in RmlUi.

Feel free to ask any questions and elaborate if you have something to bring on table.

AND I didn't get how you just annihilated CI/CD because it shows me this
<img width="1061" height="179" alt="image" src="https://github.com/user-attachments/assets/f93564b8-8622-43ae-90a6-312018743642" />

Hope you will restore it back and we will see nice blue/green ticks that show that CI/CD builts successfully.

hyblocker committed:
Merge branch 'master' into dx11-backend

Bimbol95 commented:
So will this be merged anytime in the future?

mikke89 commented:
Alright, so I just merged this one now 🎉

First of all, thanks a lot for the implementation @hyblocker, this is very much a welcome addition. It's very nice to see another fully-feature backend.

I made some smaller changes before merging, mainly:

- Tried to make it easier to inject users' own buffers/render targets: `BeginFrame()` no longer takes any arguments and no longer automatically clears. Instead, added a separate `Clear()` function so users can choose when to clear their own buffers. `EndFrame` now takes a target view to draw the final result to.
- Fix an issue when minimizing the window, now handles a zero-sized window.
- Fix a resizing issue on the `GLFW_DX11` backend.
- Format according to our clang-format, various code cleanup.
- Add backends to CI.

Otherwise, some open issues remain, particularly ones mentioned here:
- https://github.com/mikke89/RmlUi/pull/675#issuecomment-2821714716

There is still room for performance and other improvements, but not considered critical, this applies to several of our renderers. With this merged, I hope it can get more exposure and more contributions.

Thanks again!
